pkg/operator/yugabytedb/create_controller.go

```diff
@@ -0,0 +1,343 @@
+	serviceName = cluster.addCRNameSuffix(serviceName)
+	label = cluster.addCRNameSuffix(label)
+
+	// This service is meant to be used by clients of the database. It exposes a ClusterIP that will
+	// automatically load balance connections to the different database pods.
+	uiService := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      serviceName,
+			Namespace: cluster.namespace,
+			Labels:    createAppLabels(label),
+		},
+		Spec: v1.ServiceSpec{
+			Selector: createAppLabels(label),
+			Type:     v1.ServiceTypeClusterIP,
+			Ports:    createUIServicePorts(ports, isTServerService),
+		},
+	}
+	k8sutil.SetOwnerRef(&uiService.ObjectMeta, &cluster.ownerRef)
+
+	if _, err := c.context.Clientset.CoreV1().Services(cluster.namespace).Create(uiService); err != nil {
+		if !errors.IsAlreadyExists(err) {
+			return err
+		}
+		logger.Infof("client service %s already exists in namespace %s", uiService.Name, uiService.Namespace)
+	} else {
+		logger.Infof("client service %s started in namespace %s", uiService.Name, uiService.Namespace)
+	}
+
+	return nil
+}
+
+func (c *ClusterController) createMasterHeadlessService(cluster *cluster) error {
+	return c.createHeadlessService(cluster, false)
+}
+
+func (c *ClusterController) createTServerHeadlessService(cluster *cluster) error {
+	return c.createHeadlessService(cluster, true)
+}
+
+func (c *ClusterController) createHeadlessService(cluster *cluster, isTServerService bool) error {
+	serviceName := masterNamePlural
+	label := masterName
+
+	if isTServerService {
+		serviceName = tserverNamePlural
+		label = tserverName
+	}
+
+	// Append CR name suffix to make the service name & label unique in current namespace.
+	serviceName = cluster.addCRNameSuffix(serviceName)
+	label = cluster.addCRNameSuffix(label)
+
+	// This service only exists to create DNS entries for each pod in the stateful
+	// set such that they can resolve each other's IP addresses. It does not
+	// create a load-balanced ClusterIP and should not be used directly by clients
+	// in most circumstances.
+	headlessService := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      serviceName,
+			Namespace: cluster.namespace,
+			Labels:    createAppLabels(label),
+		},
+		Spec: v1.ServiceSpec{
+			Selector: createAppLabels(label),
+			// We want all pods in the StatefulSet to have their addresses published for
+			// the sake of the other YugabyteDB pods even before they're ready, since they
+			// have to be able to talk to each other in order to become ready.
+			ClusterIP: "None",
+			Ports:     createServicePorts(cluster, isTServerService),
+		},
+	}
+
+	k8sutil.SetOwnerRef(&headlessService.ObjectMeta, &cluster.ownerRef)
+
+	if _, err := c.context.Clientset.CoreV1().Services(cluster.namespace).Create(headlessService); err != nil {
+		if !errors.IsAlreadyExists(err) {
+			return err
+		}
+		logger.Infof("headless service %s already exists in namespace %s", headlessService.Name, headlessService.Namespace)
```

Comment:
Probably worth for the future to look into the "UpdateService" funcs to make sure the Service is as needed. See next comment.

Reply:
Please see next comment.